BUG: fix stacklevel in warning within random.shuffle

seberg opened the pull request:
Cython does _not_ add a new stacklevel (at least at this time),
so `stacklevel=1` (or ommiting it) is, maybe surprisingly, correct.

The second commit adds a test for the warning, with an assert for the filename (which by proxy tests the stacklevel).  I am not quite sure how safe that test is, so if we want to backport we could omit it, I guess.

seberg committed:
BUG: fix stacklevel in warning within random.shuffle

Cython does _not_ add a new stacklevel (at least at this time),
so `stacklevel=1` (or ommiting it) is, maybe surprisingly, correct.

seberg commented:
Honestly, mainly creating it to not have an incorrect stacklevel precedence. Most of the code does not use it (which is equivalent to 1).  Maybe this is actually a small annoyance with cython itself...

bashtage commented on numpy/random/_generator.pyx:
```diff
@@ -4448,8 +4448,7 @@ cdef class Generator:
                     "`x` isn't a recognized object; `shuffle` is not guaranteed "
```
While you are at it.  `x` isn't a recognized object isn't very helpful.  Would be clearer to say `x` is not a subclass of Sequence. Since the latter has meaning?

seberg replied:
good point changed it to:
```
UserWarning: you are shuffling a 'dict' object which is not a subclass of 'Sequence'; `shuffle` is not guaranteed to behave correctly. E.g., non-numpy array/tensor objects with view semantics may contain duplicates after shuffling
```

seberg committed:
TST: Add test for shuffle warning and assert stacklevel

The test checks that the warning originates in the correct file
(test_random.py). I am not quite sure how safe the test is, though.

Unfortunately, there is no "obvious" way to test stacklevels.

mattip commented:
Thanks @seberg
